flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/expressions/resolver/ExpressionResolverTest.java

```diff
@@ -0,0 +1,391 @@
+						.field("f0", DataTypes.INT())
+						.build()
+				)
+				.lookupFunction("func", new ScalarFunc())
+				.select(call("func", call(new ScalarFunc(), call("func", 1, $("f0")))))
+				.equalTo(
+					new CallExpression(
+						FunctionIdentifier.of("func"),
+						new ScalarFunc(),
+						Collections.singletonList(
+							new CallExpression(
+								new ScalarFunc(),
+								Collections.singletonList(new CallExpression(
+									FunctionIdentifier.of("func"),
+									new ScalarFunc(),
+									Arrays.asList(
+										valueLiteral(1),
+										new FieldReferenceExpression("f0", DataTypes.INT(), 0, 0)),
+									DataTypes.INT().notNull().bridgedTo(int.class)
+								)),
+								DataTypes.INT().notNull().bridgedTo(int.class)
+							)),
+						DataTypes.INT().notNull().bridgedTo(int.class))
+				)
+		);
+	}
+
+	@Parameterized.Parameter
+	public TestSpec testSpec;
+
+	@Test
+	public void testResolvingExpressions() {
+		List<ResolvedExpression> resolvedExpressions = testSpec.getResolver()
+			.resolve(Arrays.asList(testSpec.expressions));
+		assertThat(
+			resolvedExpressions,
+			equalTo(testSpec.expectedExpressions));
+	}
+
+	/**
+	 * Test scalar function.
+	 */
+	@FunctionHint(input = @DataTypeHint(inputGroup = InputGroup.ANY), isVarArgs = true, output = @DataTypeHint(value = "INTEGER NOT NULL", bridgedTo = int.class))
+	public static class ScalarFunc extends ScalarFunction {
+		public int eval(Object... any) {
+			return 0;
+		}
+
+		@Override
+		public int hashCode() {
+			return 0;
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			return obj instanceof ScalarFunc;
+		}
+	}
+
+	private static class TestSpec {
+		private final String description;
+		private TableSchema[] schemas;
+		private Expression[] expressions;
+		private List<ResolvedExpression> expectedExpressions;
+		private Map<FunctionIdentifier, FunctionDefinition> functions = new HashMap<>();
+
+		private TestSpec(String description) {
+			this.description = description;
+		}
+
+		public static TestSpec test(String description) {
+			return new TestSpec(description);
+		}
+
+		public TestSpec inputSchemas(TableSchema... schemas) {
+			this.schemas = schemas;
+			return this;
+		}
+
+		public TestSpec lookupFunction(String name, FunctionDefinition functionDefinition) {
+			functions.put(FunctionIdentifier.of(name), functionDefinition);
+			return this;
+		}
+
+		public TestSpec lookupFunction(ObjectIdentifier identifier, FunctionDefinition functionDefinition) {
+			functions.put(FunctionIdentifier.of(identifier), functionDefinition);
+			return this;
+		}
+
+		public TestSpec select(Expression... expressions) {
+			this.expressions = expressions;
+			return this;
+		}
+
+		public TestSpec equalTo(ResolvedExpression... resolvedExpressions) {
+			this.expectedExpressions = Arrays.asList(resolvedExpressions);
+			return this;
+		}
+
+		public ExpressionResolver getResolver() {
+			DataTypeFactory dataTypeFactory = new DataTypeFactory() {
```

Comment:
Use `org.apache.flink.table.types.inference.utils.DataTypeFactoryMock` I already moved it to a more prominent location in my other PR.